nitro-root-enclave/src/main.rs

```diff
@@ -0,0 +1,467 @@
+    0xd2, 0x55, 0x03, 0x8e, 0x1f, 0x03, 0x43, 0xf9, 0x5b, 0x63, 0xa9, 0x62, 0x8c, 0x3d, 0x39, 0x80,
+    0x95, 0x45, 0xa1, 0x1e, 0xbc, 0xbf, 0x2e, 0x3b, 0x55, 0xd8, 0xae, 0xee, 0x71, 0xb4, 0xc3, 0xd6,
+    0xad, 0xf3, 0x02, 0x31, 0x00, 0xa2, 0xf3, 0x9b, 0x16, 0x05, 0xb2, 0x70, 0x28, 0xa5, 0xdd, 0x4b,
+    0xa0, 0x69, 0xb5, 0x01, 0x6e, 0x65, 0xb4, 0xfb, 0xde, 0x8f, 0xe0, 0x06, 0x1d, 0x6a, 0x53, 0x19,
+    0x7f, 0x9c, 0xda, 0xf5, 0xd9, 0x43, 0xbc, 0x61, 0xfc, 0x2b, 0xeb, 0x03, 0xcb, 0x6f, 0xee, 0x8d,
+    0x23, 0x02, 0xf3, 0xdf, 0xf6,
+];
+
+// the following value was copied from https://github.com/aws/aws-nitro-enclaves-sdk-c/blob/main/source/attestation.c
+// I've no idea where it came from (I've seen no documentation on this), but
+// I guess I have to trust Amazon on this one
+const NSM_MAX_ATTESTATION_DOC_SIZE: usize = 16 * 1024;
+
+lazy_static! {
+    static ref DEVICE_KEY_PAIR: Mutex<Option<(EcdsaKeyPair, Vec<u8>)>> = Mutex::new(None);
+    static ref MEXICO_CITY_HASH: Mutex<Option<Vec<u8>>> = Mutex::new(None);
+    static ref DEVICE_ID: Mutex<Option<i32>> = std::sync::Mutex::new(None);
+}
+
+fn get_firmware_version() -> Result<String, String> {
+    println!("nitro-root-enclave::get_firmware_version");
+    let version = env!("CARGO_PKG_VERSION");
+    return Ok(version.to_string());
+}
+
+fn set_mexico_city_hash_hack(hash: Vec<u8>) -> Result<NitroStatus, String> {
+    let mut mch_guard = MEXICO_CITY_HASH.lock().map_err(|err| {
+        format!(
+            "set_mexico_city_hash failed to obtain lock on MEXICO_CITY_HASH:{:?}",
+            err
+        )
+    })?;
+    *mch_guard = Some(hash);
+    Ok(NitroStatus::Success)
+}
+
+fn native_attestation(challenge: &Vec<u8>, device_id: i32) -> Result<(Vec<u8>, Vec<u8>), String> {
+    let mut att_doc: Vec<u8> = vec![0; NSM_MAX_ATTESTATION_DOC_SIZE];
+
+    {
+        let mut di_guard = DEVICE_ID.lock().map_err(|err| {
+            format!(
+                "nitro-root-enclave::native_attestation failed to obtain lock on DEVICE_ID:{:?}",
+                err
+            )
+        })?;
+        *di_guard = Some(device_id);
+    }
+
+    let mut att_doc_len: u32 = att_doc.len() as u32;
+    let device_private_key = {
+        let dkp_guard = DEVICE_KEY_PAIR.lock().map_err(|err| format!("nitro-root-enclave::native_attestation failed to obtain lock on DEVICE_KEY_PAIR:{:?}", err))?;
+        match &*dkp_guard {
+            Some((_key, bytes)) => bytes.clone(),
+            None => return Err(format!("nitro-root-enclave::native_attestation for some reason the DEVICE_KEY_PAIR is uninitialized. I don't know how you got here")),
+        }
+    };
+    let device_public_key = {
+        // Oddity: the current way of extracting the public key from the private
+        // key is to load it into PSA Crytpo, and then extract the public
+        // component. There are better ways to do this, but this is what
```

Comment:
Maybe add an issue to address this?  Sounds like it might be a good first issue!

Reply:
I agree we should file an issue, but I don't think it would be a good first issue. Key formats are a horrible mess. I've made a lot of money in my career fighting with differing key format problems.

Issue #65 opened.